content/en/platform/corda/5.0/developing-applications/getting-started/utxo-advanced-ledger-extensions/commands-example.md

```diff
@@ -0,0 +1,243 @@
+
+{{< note >}}
+There are cases where this design approach no longer fits the design goals of the system being implemented. Specifically, in regard to contract extensibility, it is currently not possible to extend a contract to support additional constraints.
+{{< /note >}}
+
+## Derivable Contract Design
+
+The following contract refactors the above to support the ability to derive contracts, and provide additional constraints in a secure and controlled way.
+The contract still provides the same three commands; `Create`, `Update` and `Delete`. The verify function delegates these command types to `verifyCreate`, `verifyUpdate` and `verifyDelete` functions respectively, which in turn call `onVerifyCreate`, `onVerifyUpdate` and `onVerifyDelete` respectively.
+
+{{< note >}}
+The verify function has been marked final. This change is necessary as it prevents derived contract implementations from circumventing the base contract rules.
+{{< /note >}}
+
+  ```kotlin
+  public class ExampleContract implements Contract {
+  
+    private interface ExampleContractCommand extends Command { }
+    public static class Create implements ExampleContractCommand { }
+    public static class Update implements ExampleContractCommand { }
+    public static class Delete implements ExampleContractCommand { }
+    
+    @Override
+    
+    public final void verify(UtxoLedgerTransaction transaction) {
+    
+        List<? extends ExampleContractCommand> commands = transaction
+                .getCommands(ExampleContractCommand.class);
+                
+        for (ExampleContractCommand command : commands) {
+            if (command instanceof Create) verifyCreate(transaction);
+            else if (command instanceof Update) verifyUpdate(transaction);
+            else if (command instanceof Delete) verifyDelete(transaction);
+            else throw new IllegalStateException("Unrecognised command type.");
+        }
+    }
+    
+    protected void onVerifyCreate(UtxoLedgerTransaction transaction) { }
+    protected void onVerifyUpdate(UtxoLedgerTransaction transaction) { }
+    protected void onVerifyDelete(UtxoLedgerTransaction transaction) { }    
+    private void verifyCreate(UtxoLedgerTransaction transaction) {
+    
+        // Verify base Create constraints
+        // Then verify additional Create constraints implemented by derived contracts
+        onVerifyCreate(transaction);
+        
+    }
+    private void verifyUpdate(UtxoLedgerTransaction transaction) {
+        // Verify base Update constraints
+        // Then verify additional Update constraints implemented by derived contracts
+        onVerifyUpdate(transaction);
+    }
+    
+    private void verifyDelete(UtxoLedgerTransaction transaction) {
+        // Verify base Delete constraints
+        // Then verify additional Delete constraints implemented by derived contracts
+        onVerifyDelete(transaction);
+    }
+}
+```
```

Comment:
```suggestion
```kotlin
public class ExampleContract implements Contract {
 
    private interface ExampleContractCommand extends Command { }

    public static class Create implements ExampleContractCommand { }
    public static class Update implements ExampleContractCommand { }
    public static class Delete implements ExampleContractCommand { }
    
    @Override
    public final void verify(UtxoLedgerTransaction transaction) {
        List<? extends ExampleContractCommand> commands = transaction
                .getCommands(ExampleContractCommand.class);
                
        for (ExampleContractCommand command : commands) {
            if (command instanceof Create) verifyCreate(transaction);
            else if (command instanceof Update) verifyUpdate(transaction);
            else if (command instanceof Delete) verifyDelete(transaction);
            else throw new IllegalStateException("Unrecognised command type.");
        }
    }
    
    protected void onVerifyCreate(UtxoLedgerTransaction transaction) { }
    protected void onVerifyUpdate(UtxoLedgerTransaction transaction) { }
    protected void onVerifyDelete(UtxoLedgerTransaction transaction) { }    

    private void verifyCreate(UtxoLedgerTransaction transaction) {
        // Verify base Create constraints
        // Then verify additional Create constraints implemented by derived contracts
        onVerifyCreate(transaction);
    }
    
    private void verifyUpdate(UtxoLedgerTransaction transaction) {
        // Verify base Update constraints
        // Then verify additional Update constraints implemented by derived contracts
        onVerifyUpdate(transaction);
    }
    
    private void verifyDelete(UtxoLedgerTransaction transaction) {
        // Verify base Delete constraints
        // Then verify additional Delete constraints implemented by derived contracts
        onVerifyDelete(transaction);
    }
}
```
```